rfcs/text/0013_background_sessions.md

```diff
@@ -0,0 +1,484 @@
+
+   /**
+    * Extend a session and all associated searchIds.
+    * @param request 
+    * @param sessionId 
+    * @param extendBy Time to extend by, can be a relative or absolute string.
+    * @returns success status
+    * @throws Throws an error in OSS.
+    */
+  extend: async (
+    request: KibanaRequest,
+    sessionId: string,
+    extendBy: string,
+  ) => Promise<boolean>
+
+   /**
+    * Get a list of background session objects.
+    * @param request 
+    * @param sessionId
+    * @returns success status
+    * @throws Throws an error in OSS.
+    */
+  list: async (
+    request: KibanaRequest,
+  ) => Promise<BackgroundSessionAttributes[]>
+
+   /**
+    * Update the status of a given session
+    * @param request 
+    * @param sessionId
+    * @param status
+    * @returns success status
+    * @throws Throws an error in OSS.
+    */
+  updateStatus: async (
+    request: KibanaRequest,
+    sessionId: string,
+    status: BackgroundSessionStatus
+  ) => Promise<boolean>
+}
+
+```
+
+### Search Service Changes
+
+There are cases where search requests are issued by the server (Like TSVB).
+We can simplify this flow by intoducing a mechanism, similar to the frontend one, tracking the information in memory. 
+
+However, due to the fact that we are running in a potenitally multi-server architecture, each server will have to poll for a saved object with a corresponding sessionId to store the ids into it.
+
+```ts
+interface SearchService {
+  /**
+   * The search API will accept the option `trackId`, which will track the search ID, if available, on the server, until a corresponding saved object is created.  
+   **/
+  search: async (
+    context: RequestHandlerContext,	  
+    request: IEnhancedEsSearchRequest,
+    options?: ISearchOptions
+  ) => ISearchResponse<Payload=any>
+}
+```
+
+### Server Routes
+
+Each route exposes the corresponding method from the Session Service
+
+`POST /internal/session/store`
+
+`POST /internal/session/extend`
+
+`POST /internal/session/expire`
+
+`GET /internal/session/list`
+
+### Search Strategy Integration 
+
+If the `EnhancedEsSearchStrategy` receives a `restore` option, it will attempt reloading data using the Background Session saved object matching the provided `sessionId`. If there are any errors during that process, the strategy will return an error response and *not attempt to re-run the request.
+
+The strategy will track the asyncId on the server side, if `trackId` option is provided.
+
+### Monitoring Service
+
+The `data` plugin will register a task with the task manager, periodically monitoring the status of incomplete background sessions. 
+
+It will query the list of all incomplete sessions, and check the status of each search that is executing. If the search requests are all complete, it will update the corresponding saved object to have a `status` of `complete`. If any of the searches return an error, it will update the saved object to an `error` state.
```

Comment:
Should this service also be responsible for marking sessions as expired, and/or cleaning up expired sessions?

Reply:
++ Yes, I can add that to the text here.

Reply:
@lizozom Any thoughts about how often we may want to automatically "clean up" any expired background session objects?

Reply:
I would make this configurable via advanced settings (EXPIRED_SESSION_TTL?)